modules/apps/portal/portal-cluster-multiple/src/main/java/com/liferay/portal/cluster/multiple/internal/ClusterExecutorImpl.java

```diff
@@ -222,19 +203,30 @@ protected void activate(ComponentContext componentContext) {
 			ClusterExecutorConfiguration.class,
 			componentContext.getProperties());
 
+		_bundleContext = componentContext.getBundleContext();
+
 		initialize(
 			_props.get(PropsKeys.CLUSTER_LINK_CHANNEL_LOGIC_NAME_CONTROL),
 			_props.get(PropsKeys.CLUSTER_LINK_CHANNEL_PROPERTIES_CONTROL),
 			_props.get(PropsKeys.CLUSTER_LINK_CHANNEL_NAME_CONTROL));
 
-		BundleContext bundleContext = componentContext.getBundleContext();
-
-		_serviceRegistration = bundleContext.registerService(
+		_serviceRegistration = _bundleContext.registerService(
 			PortalInetSocketAddressEventListener.class,
 			new ClusterExecutorPortalInetSocketAddressEventListener(),
 			new HashMapDictionary<String, Object>());
 	}
 
+	@Reference(
+		cardinality = ReferenceCardinality.MULTIPLE,
+		policy = ReferencePolicy.DYNAMIC,
+		policyOption = ReferencePolicyOption.GREEDY
+	)
+	protected void addClusterEventListener(
```

Comment:
@W2jx It's weird that you remove the public method with the reference, but add a new protected method with the same name, doing the same thing.

Please do it in this order:

1. Convert direct method references to field references;
2. Convert the `ClusterEventListener` reference to a `ServiceTrackerList`. After this, the public API still exists, but not with any `@Reference` annotation;
3. Remove the public APIs;